File: mylibrary/src/main/assets/rich_editor.js
const RE = {};

RE.currentSelection = {
    "startContainer": 0,
    "startOffset": 0,
    "endContainer": 0,
    "endOffset": 0
};

RE.editor = document.getElementById('editor');

//当前文档选择进行了改变
document.addEventListener("selectionchange", function () {

    const left = document.queryCommandState('justifyLeft');
    const center = document.queryCommandState('justifyCenter');

    if (left) {
        RE.setJustifyCenter()
        RE.setJustifyLeft()
    } else if (center) {
        RE.setJustifyLeft()
        RE.setJustifyCenter()
    } else {
        RE.setJustifyCenter()
        RE.setJustifyRight()
    }

    RE.backuprange();
});

// Event Listeners
RE.editor.addEventListener("input", function () {
    const html = RE.getHtml();
    window.location.href = "re-callback://" + encodeURI(html);
});

// Initializations
RE.callback = function (e) {
    const html = RE.getHtml();
    window.location.href = "re-callback://" + encodeURI(html);

    setTimeout(function () {
        RE.enabledEditingItems(e);
    }, 50)

}

RE.setHtml = function (contents) {
    RE.editor.innerHTML = decodeURIComponent(contents.replace(/\+/g, '%20'));
}

RE.getHtml = function () {
    return RE.editor.innerHTML;
}

RE.getText = function () {
    return RE.editor.innerText;
}

RE.setBaseTextColor = function (color) {
    RE.editor.style.color = color;
}

RE.setBaseFontSize = function (size) {
    RE.editor.style.fontSize = size;
}

RE.setPadding = function (left, top, right, bottom) {
    RE.editor.style.paddingLeft = left;
    RE.editor.style.paddingTop = top;
    RE.editor.style.paddingRight = right;
    RE.editor.style.paddingBottom = bottom;
}

RE.setBackgroundColor = function (color) {
    document.body.style.backgroundColor = color;
}

RE.setBackgroundImage = function (image) {
    RE.editor.style.backgroundImage = image;
}

RE.setWidth = function (size) {
    RE.editor.style.minWidth = size;
}

RE.setHeight = function (size) {
    RE.editor.style.height = size;
}

RE.setTextAlign = function (align) {
    RE.editor.style.textAlign = align;
}

RE.setVerticalAlign = function (align) {
    RE.editor.style.verticalAlign = align;
}

RE.setPlaceholder = function (placeholder) {
    RE.editor.setAttribute("placeholder", placeholder);
}

RE.setInputEnabled = function (inputEnabled) {
    RE.editor.contentEditable = String(inputEnabled);
}

RE.undo = function () {
    document.execCommand('undo', false, null);
}

RE.redo = function () {
    document.execCommand('redo', false, null);
}

RE.setBold = function (textBold) {
    const selection = window.getSelection();
    if (window.getSelection().toString()) {
        document.execCommand('bold', false, null)
    } else {

        if (textBold) {
            selection.focusNode.parentNode.style.fontWeight = "bold"
        } else {
            selection.focusNode.parentNode.style.fontWeight = "normal"
        }
        RE.callback()
    }
}

RE.setItalic = function (textItalic) {
    const selection = window.getSelection();
    if (window.getSelection().toString()) {
        document.execCommand('italic', false, null);
    } else {
        if (textItalic) {
            selection.focusNode.parentNode.style.fontStyle = "italic"
        } else {
            selection.focusNode.parentNode.style.fontStyle = "normal"
        }
        RE.callback()
    }
}

RE.setSubscript = function () {
    document.execCommand('subscript', false, null);
}

RE.setSuperscript = function () {
    document.execCommand('superscript', false, null);
}

RE.setStrikeThrough = function () {
    document.execCommand('strikeThrough', false, null);
}

RE.setUnderline = function (textUnderline) {
    const selection = window.getSelection();
    if (window.getSelection().toString()) {
        document.execCommand('underline', false, null);
    } else {
        if (textUnderline) {
            selection.focusNode.parentNode.style.textDecoration = "underline"
        } else {
            selection.focusNode.parentNode.style.textDecoration = "none"
        }

        RE.callback()
    }
}

RE.setBullets = function () {
    document.execCommand('insertUnorderedList', false, null);
}

RE.setNumbers = function () {
    document.execCommand('insertOrderedList', false, null);
}

RE.setTextColor = function (color) {
    const selection = window.getSelection();
    if (window.getSelection().toString()) {
        document.execCommand('foreColor', false, color);
    } else {
        selection.focusNode.parentNode.style.color = color
        RE.callback()
    }
}


RE.setTextBackgroundColor = function (color) {
//    RE.restorerange();
    document.execCommand("styleWithCSS", null, true);
    document.execCommand('hiliteColor', false, color);
    document.execCommand("styleWithCSS", null, false);
}

RE.setFontSize = function (fontSize) {
    const selection = window.getSelection();
    if (window.getSelection().toString()) {
        document.execCommand("fontSize", false, fontSize);
    } else {
        if (fontSize === "3") {
            selection.focusNode.parentNode.style.fontSize = "medium"
        } else if (fontSize === "4") {
            selection.focusNode.parentNode.style.fontSize = "large"
        } else if (fontSize === "5") {
            selection.focusNode.parentNode.style.fontSize = "x-large"
        }
        RE.callback()
    }
}

RE.setLineHeight = function (heightInPixel) {
// 设置单行行高
//            document.execCommand('formatblock', false, 'p')
//            var selectedNodes = [];
//            var sel = rangy.getSelection();
//            for (var i = 0; i < sel.rangeCount; i++) {
//                selectedNodes = selectedNodes.concat(sel.getRangeAt(i).getNodes());
//                $(selectedNodes).css("height", heightInPixel);
//            }
//            selectedElement = window.getSelection().focusNode.parentNode;
//            $(selectedElement).css("height", heightInPixel);
    var t = document.getElementById('editor');
    t.style.lineHeight = heightInPixel;
}

RE.setHeading = function (heading) {
    document.execCommand('formatBlock', false, '<h' + heading + '>');
}

RE.setIndent = function () {
    document.execCommand('indent', false, null);
}

RE.setOutdent = function () {
    document.execCommand('outdent', false, null);
}

RE.setJustifyLeft = function () {
    document.execCommand('justifyLeft', false, null);
}

RE.setJustifyCenter = function () {
    document.execCommand('justifyCenter', false, null);
}

RE.setJustifyRight = function () {
    document.execCommand('justifyRight', false, null);
}

RE.setBlockquote = function () {
    document.execCommand('formatBlock', false, '<blockquote>');
}

RE.insertImage = function (url, alt, widthPercent) {
    var html = '<img src="' + url + '" width="' + widthPercent + '%" alt="' + alt + '" />';
    RE.insertHTML(html);
}

RE.insertImageWrapWidth = function (url, alt) {
    var html = '<img src="' + url + '" alt="' + alt + '" />';
    RE.insertHTML(html);
}

RE.insertHTML = function (html) {
    RE.restorerange();
    document.execCommand('insertHTML', false, html);
}

RE.insertText = function (text) {
    document.execCommand('insertText', false, text);
}

RE.deleteOneWord = function () {
    document.execCommand('delete', false, "1");
}


RE.insertLink = function (url, title) {
    RE.restorerange();
    const sel = document.getSelection();
    if (sel.toString().length == 0) {
        document.execCommand("insertHTML", false, "<a href='" + url + "'>" + title + "</a>");
    } else if (sel.rangeCount) {
        const el = document.createElement("a");
        el.setAttribute("href", url);
        el.setAttribute("title", title);

        const range = sel.getRangeAt(0).cloneRange();
        range.surroundContents(el);
        sel.removeAllRanges();
        sel.addRange(range);
    }
    RE.callback();
}

RE.setTodo = function (text) {
    const html = '<input type="checkbox" name="' + text + '" value="' + text + '"/> &nbsp;';
    document.execCommand('insertHTML', false, html);
}

RE.backuprange = function () {
    const selection = window.getSelection();
    const rangeCount = selection.rangeCount;
    if (rangeCount > 0) {
        const range = selection.getRangeAt(0);

        const startContainer = range.startContainer;
        const startOffset = range.startOffset;
        const endContainer = range.endContainer;
        const endOffset = range.endOffset;

        RE.currentSelection = {
            "startContainer": startContainer,
            "startOffset": startOffset,
            "endContainer": endContainer,
            "endOffset": endOffset
        };
    }
}

RE.restorerange = function () {
    const selection = window.getSelection();
    selection.removeAllRanges();
    const range = document.createRange();
    range.setStart(RE.currentSelection.startContainer, RE.currentSelection.startOffset);
    range.setEnd(RE.currentSelection.endContainer, RE.currentSelection.endOffset);
    selection.addRange(range);
}


RE.queryCommandState = function (command) {
    var items = [];
    if (document.queryCommandState(command)) {
        window.location.href = "re-state://true";
    } else {
        window.location.href = "re-state://false";
    }
}

RE.enabledEditingItems = function (e) {
    const items = [];
    if (document.queryCommandState('bold')) {
        items.push('bold');
    }
    if (document.queryCommandState('italic')) {
        items.push('italic');
    }
    if (document.queryCommandState('subscript')) {
        items.push('subscript');
    }
    if (document.queryCommandState('superscript')) {
        items.push('superscript');
    }
    if (document.queryCommandState('strikeThrough')) {
        items.push('strikeThrough');
    }
    if (document.queryCommandState('underline')) {
        items.push('underline');
    }
    if (document.queryCommandState('insertOrderedList')) {
        items.push('orderedList');
    }
    if (document.queryCommandState('insertUnorderedList')) {
        items.push('unorderedList');
    }
    if (document.queryCommandState('justifyCenter')) {
        items.push('justifyCenter');
    }
    if (document.queryCommandState('justifyFull')) {
        items.push('justifyFull');
    }
    if (document.queryCommandState('justifyLeft')) {
        items.push('justifyLeft');
    }
    if (document.queryCommandState('justifyRight')) {
        items.push('justifyRight');
    }
    if (document.queryCommandState('insertHorizontalRule')) {
        items.push('horizontalRule');
    }
    var formatBlock = document.queryCommandValue('formatBlock');
    if (formatBlock.length > 0) {
        items.push(formatBlock);
    }

    reportColourAndFontSize(items);

    window.location.href = "re-state://" + encodeURI(items.join(','));
}

RE.focus = function () {
    var range = document.createRange();
    range.selectNodeContents(RE.editor);
    range.collapse(false);
    var selection = window.getSelection();
    selection.removeAllRanges();
    selection.addRange(range);
    RE.editor.focus();
}

RE.blurFocus = function () {
    RE.editor.blur();
}

RE.removeFormat = function () {
    document.execCommand('removeFormat', false, null);
}


RE.editor.addEventListener("keyup", function (e) {
    var KEY_LEFT = 37, KEY_RIGHT = 39;
    if (e.which == KEY_LEFT || e.which == KEY_RIGHT) {
        RE.enabledEditingItems(e);
    }
});


RE.editor.addEventListener("click", function () {
    RE.enabledEditingItems();
});


// 获取文字颜色和大小
function getComputedStyleProperty(el, propName) {
    if (window.getComputedStyle) {
        return window.getComputedStyle(el, null)[propName];
    } else if (el.currentStyle) {
        return el.currentStyle[propName];
    }
}


function reportColourAndFontSize(items) {
    let containerEl, sel;
    if (window.getSelection) {
        sel = window.getSelection();
        if (sel.rangeCount) {
            containerEl = sel.getRangeAt(0).commonAncestorContainer;
            // https://www.w3school.com.cn/jsref/prop_node_nodetype.asp
            //3代表是一个文本节点
            if (containerEl.nodeType == 3) {
                containerEl = containerEl.parentNode;
            }
        }
    } else if ((sel = document.selection) && sel.type != "Control") {
        containerEl = sel.createRange().parentElement();
    }

    if (containerEl) {
        var fontSize = containerEl.size;
        console.log("----" + fontSize)
        if (!fontSize) {
            fontSize = 4;
        }
        var color = getComputedStyleProperty(containerEl, "color");
        var backgroundColor = getComputedStyleProperty(containerEl, "background-color");
        items.push("textColor:" + colorRGB2Hex(color));
        items.push("textBgColor:" + colorRGB2Hex(backgroundColor));
        items.push(fontSize + "pt");
    }
}

/*RGB颜色转换为16进制*/
function colorRGB2Hex(color) {
    var rgb = color.split(',');
    var r = parseInt(rgb[0].split('(')[1]);
    var g = parseInt(rgb[1]);
    var b = parseInt(rgb[2].split(')')[0]);
    var hex = "#" + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1);
    return hex;
}

File: mylibrary/src/main/java/com/alguojian/richeditor/RichEditor.kt
package com.alguojian.richeditor

import android.R
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Build
import android.text.TextUtils
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import com.alguojian.richeditor.RichUtils.currentTime
import com.alguojian.richeditor.RichUtils.decodeResource
import com.alguojian.richeditor.RichUtils.toBase64
import com.alguojian.richeditor.RichUtils.toBitmap
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.net.URLEncoder
import java.util.*

class RichEditor @SuppressLint("SetJavaScriptEnabled") constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int
) : WebView(context, attrs, defStyleAttr) {
    enum class Type {
        BOLD, ITALIC, SUBSCRIPT, SUPERSCRIPT, STRIKETHROUGH, UNDERLINE, H1, H2, H3, H4, H5, H6, ORDEREDLIST, UNORDEREDLIST, JUSTIFYCENTER, JUSTIFYFULL, JUSTIFYLEFT, JUSTIFYRIGHT
    }

    interface OnTextChangeListener {
        fun onTextChange(text: String)
    }

    interface OnTextContentListener {
        fun onContent(content: String)
    }

    interface OnDecorationStateListener {
        fun onStateChangeListener(
            text: String,
            types: List<Type>
        )
    }

    interface AfterInitialLoadListener {
        fun onAfterInitialLoad(isReady: Boolean)
    }

    private var isReady = false
    private var mContents: String = ""
    private var mMaxHeight = -1
    private var mTextChangeListener: OnTextChangeListener? = null
    private var onTextContentListener: OnTextContentListener? = null
    private var mDecorationStateListener: OnDecorationStateListener? = null
    private var mLoadListener: AfterInitialLoadListener? = null

    @JvmOverloads
    constructor(context: Context, attrs: AttributeSet? = null) : this(
        context,
        attrs,
        R.attr.webViewStyle
    )

    private fun createWebviewClient(): EditorWebViewClient {
        return EditorWebViewClient()
    }

    fun setOnTextChangeListener(listener: OnTextChangeListener?) {
        mTextChangeListener = listener
    }

    fun setOnTextContentListener(listener: OnTextContentListener) {
        onTextContentListener = listener
    }

    fun setOnDecorationChangeListener(listener: OnDecorationStateListener) {
        mDecorationStateListener = listener
    }

    fun setOnInitialLoadListener(listener: AfterInitialLoadListener?) {
        mLoadListener = listener
    }

    fun setMaxHeight(maxHeight: Int) {
        mMaxHeight = maxHeight
    }

    private fun callback(text: String) {
        mContents = text.replaceFirst(CALLBACK_SCHEME.toRegex(), "")
        if (mTextChangeListener != null) {
            mTextChangeListener!!.onTextChange(mContents)
        }
        if (onTextContentListener != null) {
            onTextContentListener!!.onContent(getText())
        }
    }

    private fun stateCheck(text: String) {
        val state = text.replaceFirst(STATE_SCHEME.toRegex(), "").toUpperCase(Locale.ENGLISH)
        val types: MutableList<Type> = ArrayList()
        for (type in Type.values()) {
            if (TextUtils.indexOf(state, type.name) != -1) {
                types.add(type)
            }
        }
        if (mDecorationStateListener != null) {
            mDecorationStateListener!!.onStateChangeListener(state, types)
        }
    }

    private fun applyAttributes(context: Context, attrs: AttributeSet?) {
        val attrsArray = intArrayOf(
            R.attr.gravity
        )
        val ta = context.obtainStyledAttributes(attrs, attrsArray)
        val gravity = ta.getInt(0, View.NO_ID)
        when (gravity) {
            Gravity.START -> exec("javascript:RE.setTextAlign(\"left\")")
            Gravity.END -> exec("javascript:RE.setTextAlign(\"right\")")
            Gravity.TOP -> exec("javascript:RE.setVerticalAlign(\"top\")")
            Gravity.BOTTOM -> exec("javascript:RE.setVerticalAlign(\"bottom\")")
            Gravity.CENTER_VERTICAL -> exec("javascript:RE.setVerticalAlign(\"middle\")")
            Gravity.CENTER_HORIZONTAL -> exec("javascript:RE.setTextAlign(\"center\")")
            Gravity.CENTER -> {
                exec("javascript:RE.setVerticalAlign(\"middle\")")
                exec("javascript:RE.setTextAlign(\"center\")")
            }
        }
        ta.recycle()
    }

    // No handling
    var html: String?
        get() = mContents
        set(contents) {
            var contents = contents
            if (contents == null) {
                contents = ""
            }
            try {
                exec("javascript:RE.setHtml('" + URLEncoder.encode(contents, "UTF-8") + "');")
            } catch (e: UnsupportedEncodingException) {
                // No handling
            }
            mContents = contents
        }

    fun setEditorFontColor(color: Int) {
        val hex = convertHexColorString(color)
        exec("javascript:RE.setBaseTextColor('$hex');")
    }

    fun scrollToBottom() {
        scrollTo(0, computeVerticalScrollRange())
    }

    fun setEditorFontSize(px: Int) {
        exec("javascript:RE.setBaseFontSize('" + px + "px');")
    }

    override fun setPadding(left: Int, top: Int, right: Int, bottom: Int) {
        super.setPadding(left, top, right, bottom)
        exec(
            "javascript:RE.setPadding('" + left + "px', '" + top + "px', '" + right + "px', '" + bottom
                    + "px');"
        )
    }

    override fun setPaddingRelative(start: Int, top: Int, end: Int, bottom: Int) {
        // still not support RTL.
        setPadding(start, top, end, bottom)
    }

    fun setEditorBackgroundColor(color: Int) {
        setBackgroundColor(color)
    }

    override fun setBackgroundColor(color: Int) {
        super.setBackgroundColor(color)
    }

    override fun setBackgroundResource(resid: Int) {
        val bitmap = decodeResource(context, resid)
        val base64 = toBase64(bitmap)
        bitmap.recycle()
        exec("javascript:RE.setBackgroundImage('url(data:image/png;base64,$base64)');")
    }

    override fun setBackground(background: Drawable) {
        val bitmap = toBitmap(background)
        val base64 = toBase64(bitmap)
        bitmap.recycle()
        exec("javascript:RE.setBackgroundImage('url(data:image/png;base64,$base64)');")
    }

    fun setBackground(url: String) {
        exec("javascript:RE.setBackgroundImage('url($url)');")
    }

    fun setEditorWidth(px: Int) {
        exec("javascript:RE.setWidth('" + px + "px');")
    }

    fun setEditorHeight(px: Int) {
        exec("javascript:RE.setHeight('" + px + "px');")
    }

    fun setPlaceholder(placeholder: String) {
        exec("javascript:RE.setPlaceholder('$placeholder');")
    }

    fun setInputEnabled(inputEnabled: Boolean) {
        exec("javascript:RE.setInputEnabled($inputEnabled)")
    }

    fun loadCSS(cssFile: String) {
        val jsCSSImport = "(function() {" +
                "    var head  = document.getElementsByTagName(\"head\")[0];" +
                "    var link  = document.createElement(\"link\");" +
                "    link.rel  = \"stylesheet\";" +
                "    link.type = \"text/css\";" +
                "    link.href = \"" + cssFile + "\";" +
                "    link.media = \"all\";" +
                "    head.appendChild(link);" +
                "}) ();"
        exec("javascript:$jsCSSImport")
    }

    private fun getText(): String {
        if (!html.isNullOrEmpty()) {
            val doc = Jsoup.parse(html);
            return doc.body().text();
        }
        return "";
    }

    fun queryCommandState(command: String) {
        exec("javascript:RE.queryCommandState('$command');")
    }

    fun undo() {
        exec("javascript:RE.undo();")
    }

    fun redo() {
        exec("javascript:RE.redo();")
    }

    fun setBold(isBold: Boolean) {
        exec("javascript:RE.setBold($isBold);")
    }

    fun setItalic(isItalic: Boolean) {
        exec("javascript:RE.setItalic($isItalic);")
    }

    fun setSubscript() {
        exec("javascript:RE.setSubscript();")
    }

    fun setSuperscript() {
        exec("javascript:RE.setSuperscript();")
    }

    fun setStrikeThrough() {
        exec("javascript:RE.setStrikeThrough();")
    }

    fun setUnderline(isUnderline: Boolean) {
        exec("javascript:RE.setUnderline($isUnderline);")
    }

    fun setLineHeight(heightInPixel: Int) {
        exec("javascript:RE.setLineHeight('" + heightInPixel + "px');")
    }

    fun setTextColor(color: Int) {
        exec("javascript:RE.backuprange();")
        val hex = convertHexColorString(color)
        exec("javascript:RE.setTextColor('$hex');")
    }

    fun setTextBackgroundColor(color: Int) {
        exec("javascript:RE.backuprange();")
        val hex = convertHexColorString(color)
        exec("javascript:RE.setTextBackgroundColor('$hex');")
    }

    fun setFontSize(fontSize: Int) {
        if (fontSize > 7 || fontSize < 1) {
            Log.e("RichEditor", "Font size should have a value between 1-7")
        }
        exec("javascript:RE.setFontSize('$fontSize');")
    }

    fun removeFormat() {
        exec("javascript:RE.removeFormat();")
    }

    fun setHeading(heading: Int) {
        exec("javascript:RE.setHeading('$heading');")
    }

    fun setIndent() {
        exec("javascript:RE.setIndent();")
    }

    fun setOutdent() {
        exec("javascript:RE.setOutdent();")
    }

    fun setAlignLeft() {
        exec("javascript:RE.setJustifyLeft();")
    }

    fun setAlignCenter() {
        exec("javascript:RE.setJustifyCenter();")
    }

    fun setAlignRight() {
        exec("javascript:RE.setJustifyRight();")
    }

    fun setBlockquote() {
        exec("javascript:RE.setBlockquote();")
    }

    fun setBullets() {
        exec("javascript:RE.setBullets();")
    }

    fun setNumbers() {
        exec("javascript:RE.setNumbers();")
    }

    fun insertText(text: String) {
        exec("javascript:RE.insertText($text);")
    }

    fun deleteOneWord() {
        exec("javascript:RE.deleteOneWord();")
    }

    fun insertImage(url: String, alt: String, imageWidthPercent: Int) {
        exec("javascript:RE.backuprange();")
        exec("javascript:RE.insertImage('$url', '$alt','$imageWidthPercent');")
    }

    fun insertImageWrapWidth(url: String, alt: String) {
        exec("javascript:RE.backuprange();")
        exec("javascript:RE.insertImageWrapWidth('$url', '$alt');")
    }

    fun insertLink(href: String, title: String) {
        exec("javascript:RE.backuprange();")
        exec("javascript:RE.insertLink('$href', '$title');")
    }

    fun insertTodo() {
        exec("javascript:RE.backuprange();")
        exec("javascript:RE.setTodo('$currentTime');")
    }

    fun focusEditor() {
        requestFocus()
        exec("javascript:RE.focus();")
    }

    fun refreshState() {
        exec("javascript:RE.enabledEditingItems();")
    }

    fun clearFocusEditor() {
        exec("javascript:RE.blurFocus();")
    }

    private fun convertHexColorString(color: Int): String {
        return String.format("#%06X", 0xFFFFFF and color)
    }

    private fun exec(trigger: String) {
        if (isReady) {
            load(trigger)
        } else {
            postDelayed({ exec(trigger) }, 100)
        }
    }

    private fun load(trigger: String) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            evaluateJavascript(trigger, null)
        } else {
            loadUrl(trigger)
        }
    }

    private inner class EditorWebViewClient : WebViewClient() {
        override fun onPageFinished(view: WebView, url: String) {
            isReady = url.equals(SETUP_HTML, ignoreCase = true)
            if (mLoadListener != null) {
                mLoadListener!!.onAfterInitialLoad(isReady)
            }
        }

        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            val decode: String
            decode = try {
                URLDecoder.decode(url, "UTF-8")
            } catch (e: UnsupportedEncodingException) {
                return false
            }
            if (TextUtils.indexOf(url, CALLBACK_SCHEME) == 0) {
                callback(decode)
                return true

                //字符串中第一次出现子字符串的字符位置,标识是否已什么开头
            } else if (TextUtils.indexOf(url, STATE_SCHEME) == 0) {
                stateCheck(decode)
                return true
            }
            return super.shouldOverrideUrlLoading(view, url)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        if (mMaxHeight > -1 && measuredHeight > mMaxHeight) {
            setMeasuredDimension(measuredWidth, mMaxHeight)
        }
    }

    private var clickTime: Long = 0
    private var clickCount = 0
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_UP -> {
                if (System.currentTimeMillis() - clickTime < 500) {
                    clickCount++
                } else {
                    clickCount = 1
                }
                clickTime = System.currentTimeMillis()
                canUpdate = clickCount < 2
            }
        }
        return super.onTouchEvent(event)
    }

    var canUpdate = true

    companion object {
        private const val SETUP_HTML = "file:///android_asset/editor.html"
        private const val CALLBACK_SCHEME = "re-callback://"
        private const val STATE_SCHEME = "re-state://"
    }

    init {
        isVerticalScrollBarEnabled = false
        isHorizontalScrollBarEnabled = false
        settings.javaScriptEnabled = true
        webChromeClient = WebChromeClient()
        webViewClient = createWebviewClient()
        loadUrl(SETUP_HTML)
        applyAttributes(context, attrs)
    }
}